### PR TITLE
Add support for aggregations over distinct inputs to StreamingAggregation

### LINKED ISSUE #7666: Add support for aggregations over unique inputs to StreamingAggregation
### Description

StreamingAggregation operator currently doesn't support aggregating over unique inputs. It would be nice to add this capability. This logic exists in HashAggregation and uses DistinctAggregations helper class.

CC: @aditi-pandit @duanmeng @xiaoxmeng

### PR DESCRIPTION
**Pull request by duanmeng**
Fixes #7666

### CODE REVIEW

File: velox/exec/StreamingAggregation.cpp [hunk 1]
```diff
@@ -104,11 +115,15 @@ bool equalKeys(
 } // namespace
 
 char* StreamingAggregation::startNewGroup(vector_size_t index) {
-  if (numGroups_ < groups_.size()) {
-    auto group = groups_[numGroups_++];
-    rows_->initializeRow(group, true);
-    storeKeys(group, index);
-    return group;
+  // TODO: Fix the issue that RowContainer::initializeRow would failed when
```
Comment (mbasmanova) [inline velox/exec/StreamingAggregation.cpp:118]: Is there a GitHub issue that describes this problem? Should we fix it first before introducing this PR?

File: velox/exec/StreamingAggregation.cpp [hunk 2]
```diff
@@ -318,27 +355,32 @@ void StreamingAggregation::initializeNewGroups(size_t numPrevGroups) {
...
 
 void StreamingAggregation::initializeAggregates(uint32_t numKeys) {
-  for (auto i = 0; i < aggregates_.size(); ++i) {
-    auto& function = aggregates_[i].function;
+  for (auto& aggregate : aggregates_) {
+    auto& function = aggregate.function;
     function->setAllocator(&rows_->stringAllocator());
 
-    const auto rowColumn = rows_->columnAt(numKeys + i);
+    const auto rowColumn = rows_->columnAt(numKeys++);
```
Comment (mbasmanova) [inline velox/exec/StreamingAggregation.cpp:383]: modifying input arguments is an anti pattern; let's not do that

**Review by mbasmanova — COMMENTED** [review]
@duanmeng Thank you for adding this functionality. Have a question below and also wonder if it makes sense to extend Aggregation Fuzzer to cover this use case.

File: velox/exec/StreamingAggregation.cpp (hunk 1 above)
Comment (duanmeng) [reply velox/exec/StreamingAggregation.cpp:118]: @mbasmanova Hi Masha, thanks for your review and guidance.
>Is there a GitHub issue that describes this problem?

No, this only happens when enabling disaggregation and the **reuse flag** is true in this PR that the `accumulator.destroy(rows)` in `RowContainer::freeAggregates` would throw an error,
```C++
E20240103 08:39:44.107774 3532646 Exceptions.h:68] Line: /Users/macduan/work/native/velox-8/velox/common/memory/HashStringAllocator.cpp:426, Function:free, Expression: !header->isFree() , Source: RUNTIME, ErrorCode: INVALID_STATE
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Retriable: False
Expression: !header->isFree()
Function: free
File: /Users/macduan/work/native/velox-8/velox/common/memory/HashStringAllocator.cpp
```

>Should we fix it first before introducing this PR?

OK, I will keep investigating this issue first.

Comment (duanmeng) [reply velox/exec/StreamingAggregation.cpp]: @mbasmanova Hi Masha, I found the root cause of this failure. I've updated the PR to fix it. Could you please help take a look? Thanks.

- The `DistinctAggregations::extractValues` calls the `aggregate.function->destroy(groups)` to destroy the groups at the end, which would call **`ArrayAggregate::desctroy`**.
- The `StreamingAggregation::startNewGroup` calls `RowContainer::initializeRow`, which calls `RowContainer::freeAggregates`. `RowContainer::freeAggregates` loops the `RowContainer::accumulators_`, and calls `accumulator.destroy(rows)` , which would call **`ArrayAggregate::desctroy`** too.
- The same memory would be freed twice, and cause the error.

```C++
// DistinctAggregations::extractValues
void extractValues(folly::Range<char**> groups, const RowVectorPtr& result)
      override {
    ...
      // Release memory back to HashStringAllocator to allow next
      // aggregate to re-use it.
      aggregate.function->destroy(groups);
    }
  }

char* StreamingAggregation::startNewGroup(vector_size_t index) {
  if (numGroups_ < groups_.size()) {
    auto group = groups_[numGroups_++];
    rows_->initializeRow(group, true);

// RowContainer::initializeRow
char* RowContainer::initializeRow(char* row, bool reuse) {
  if (reuse) {
    auto rows = folly::Range<char**>(&row, 1);
    freeVariableWidthFields(rows);
    freeAggregates(rows);

void RowContainer::freeAggregates(folly::Range<char**> rows) {
  for (auto& accumulator : accumulators_) {
    accumulator.destroy(rows);
  }
}

// ArrayAggregate::desctroy
void destroy(folly::Range<char**> groups) override {
    for (auto group : groups) {
      value<ArrayAccumulator>(group)->elements.free(allocator_);
    }
  }

```

File: velox/exec/StreamingAggregation.cpp (hunk 2 above)
Comment (duanmeng) [reply velox/exec/StreamingAggregation.cpp]: Fixed

File: velox/exec/StreamingAggregation.cpp (hunk 1 above)
Comment (mbasmanova) [reply velox/exec/StreamingAggregation.cpp]: Here is how we deal with the same issue in SortedAggregations. Perhaps, same "fix" can be applied to DistinctAggregations::extractValues.

```
void SortedAggregations::extractValues(
    folly::Range<char**> groups,
    const RowVectorPtr& result) {
  raw_vector<int32_t> temp;
...
    for (const auto& aggregate : aggregates) {
      aggregate->function->extractValues(
          groups.data(), groups.size(), &result->childAt(aggregate->output));

      // Release memory back to HashStringAllocator to allow next aggregate to
      // re-use it.
      aggregate->function->destroy(groups);
      // Overwrite empty groups over the destructed groups to keep the container
      // in a well formed state.
      aggregate->function->initializeNewGroups(
          groups.data(),
          folly::Range<const int32_t*>(
              iota(groups.size(), temp), groups.size()));
    }
```

Comment (duanmeng) [reply velox/exec/StreamingAggregation.cpp]: It works, thanks for your help.

**Comment by mbasmanova** [comment]
@duanmeng Would you add "Fixes #xxx" comment to PR description to link this PR to the GitHub issue and allow that issue to auto-close when PR is merged?

File: velox/exec/tests/StreamingAggregationTest.cpp [hunk 3]
```diff
@@ -146,6 +146,43 @@ class StreamingAggregationTest : public OperatorTestBase {
             "SELECT c0, max(c1 order by c2), max(c1 order by c2 desc), array_agg(c1 order by c2) FROM tmp GROUP BY c0");
   }
 
+  void testDistinctAggregation(
+      const std::vector<VectorPtr>& keys,
+      uint32_t outputBatchSize = 1'024) {
```
Comment (mbasmanova) [inline velox/exec/tests/StreamingAggregationTest.cpp:151]: perhaps, remove default value to make it easier to read the code (no need to guess what the value is)

File: velox/exec/tests/StreamingAggregationTest.cpp [hunk 4]
```diff
@@ -385,22 +422,17 @@ TEST_F(StreamingAggregationTest, sortedAggregations) {
 }
 
 TEST_F(StreamingAggregationTest, distinctAggregations) {
-  auto data = makeRowVector({
-      makeFlatVector<int64_t>(1'000, [](auto row) { return row / 4; }),
-      makeFlatVector<int64_t>(1'000, [](auto row) { return row % 3; }),
-  });
+  auto size = 128;
```
Comment (mbasmanova) [inline velox/exec/tests/StreamingAggregationTest.cpp:425]: Why 123 and not say 1'000?

File: velox/exec/tests/StreamingAggregationTest.cpp [hunk 5]
```diff
@@ -385,22 +422,17 @@ TEST_F(StreamingAggregationTest, sortedAggregations) {
...
-                  .values({data})
-                  .streamingAggregation(
-                      {"c0"},
-                      {"array_agg(distinct c1)"},
-                      {},
-                      core::AggregationNode::Step::kSingle,
-                      false)
-                  .planNode();
+  std::vector<VectorPtr> keys = {
+      makeFlatVector<int32_t>(size, [](auto row) { return row; }),
```
Comment (mbasmanova) [inline velox/exec/tests/StreamingAggregationTest.cpp:428]: Are we testing with single-row groups only? Is this intentional?

**Review by mbasmanova — COMMENTED** [review]
@duanmeng The code changes looks good to me, but have some questions about testing. Also, still "wonder if it makes sense to extend Aggregation Fuzzer to cover this use case."

File: velox/exec/tests/StreamingAggregationTest.cpp (hunk 4 above)
Comment (duanmeng) [reply velox/exec/tests/StreamingAggregationTest.cpp:425]: 128 & 1024 are the same scenario as they are all less or equal to the default output batch (1024, `testDistinctAggregation(keys)`) and greater than the smaller output batch (32, `testDistinctAggregation(keys, 32)`). So I used a smaller size the make the test run faster.

File: velox/exec/tests/StreamingAggregationTest.cpp (hunk 5 above)
Comment (duanmeng) [reply velox/exec/tests/StreamingAggregationTest.cpp:428]: No, I just copied it from other tests. I will add multikey test cases.

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, I've resolved the latest comments and added the streaming distinct fuzzer test along with its stats. Could you help to take another look? Thanks.

File: velox/exec/fuzzer/AggregationFuzzer.cpp [hunk 6]
```diff
@@ -84,6 +84,9 @@ class AggregationFuzzer : public AggregationFuzzerBase {
     // Number of iterations using aggregations over distinct inputs.
     size_t numDistinctInputs{0};
 
+    // Number of iterations using streaming aggregations over distinct inputs.
+    size_t numStreamingDistinctInputs{0};
```
Comment (mbasmanova) [inline velox/exec/fuzzer/AggregationFuzzer.cpp:88]: This new counter seems to introduce inconsistence. We could streaming aggs with distinct inputs, but not with inputs as is. I suggest to remove this counter for now or figure out how to add it in a consistent way.

Comment (duanmeng) [reply velox/exec/fuzzer/AggregationFuzzer.cpp]: Got it, removed. I will figure it out and add the related stats separately.

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, I've rebased and added a minor optimization by removing the unnecessary `function->extractValues` of the distinct aggregation in https://github.com/facebookincubator/velox/pull/8176/commits/e9234a735e6b13de8196c19aa693296033e6308b. Could you please help take a look? Thanks a lot.

File: velox/exec/tests/StreamingAggregationTest.cpp [hunk 7]
```diff
@@ -224,6 +278,84 @@ class StreamingAggregationTest : public OperatorTestBase {
 
     EXPECT_EQ(NonPODInt64::constructed, NonPODInt64::destructed);
   }
+
+  void testMultiKeyDistinctAggregation(
```
Comment (mbasmanova) [inline velox/exec/tests/StreamingAggregationTest.cpp:282]: This tests doesn't seem to use streaming aggregation. Is this intentional?

**Review by mbasmanova — COMMENTED** [review]
@duanmeng Looks great % one question about the test.

File: velox/exec/tests/StreamingAggregationTest.cpp (hunk 7 above)
Comment (duanmeng) [reply velox/exec/tests/StreamingAggregationTest.cpp:282]: No, Fixed. I should use `PlanBuilder::streamingAggregation` explicitly and use `AggregationNode::Step::kSingle`.

**Review by mbasmanova — APPROVED** [review]
Thanks.

**Pull request closed without merging**